A self-describing representation for function closures

xavierleroy opened the pull request:
(This PR is a resurrect of #1156 following the suggestions made in comment https://github.com/ocaml/ocaml/pull/1156#pullrequestreview-35788965.)

## The context

Today, a closure for a function or a set of mutually-recursive functions is a heap block with the following structure:
```
closure ::= entrypoint (infix-header entrypoint)* value*
```
The values are the environment part of the closure: the values of the free variables.  Each entrypoint is a single code pointer in bytecode, and a 2- or 3- word record in native code:
```
entrypoint ::= code-pointer                         (in bytecode)

entrypoint ::= code-pointer  arity=1                (in native code)
             | code-pointer  arity<>1  code-pointer
```
Note that the closure really consists in two consecutive parts:
- the entrypoints part, comprising code pointers and tagged integers (infix headers look like tagged integers)
- the environment part, comprising well-formed OCaml values.

## The problem

Given a closure block at run-time, there is no way to determine which well-formed OCaml values it contains.  More precisely, there is no way to know where the entrypoints end and where the environment starts, since the first value of the environment can be a tagged integer that looks like an infix header.

This is problematic for the GC in particular, which needs to follow OCaml values that are pointers into the OCaml heap, but must not follow code pointers.  Today, we use a page table to distinguish the two kinds of pointers.  But we want to get rid of the page table (the "no naked pointers" mode), for better performance and for Multicore OCaml compatibility.

## The solution

The solution implemented in this PR adds to each "entrypoint" record the position of the environment, i.e. the field number for the first value of the value sequence that represents the environment.

In native code we just steal some bits from the arity field:
```
entrypoint ::= code-pointer closure-info                (with arity = 1)
             | code-pointer closure-info code-pointer

closure-info ::=  arity (8 bits) . start-of-environment (wordsize - 9 bits) . 1
```
(#9620 ensures that ocamlopt never generates functions with arities that would not fit in 8 bits.)

In bytecode, we need to add a closure-info word to each entrypoint, making closures a bit larger:
```
entrypoint ::= code-pointer closure-info                (with arity = 0)
```
(The '0' value for the arity field is arbitrary, as the abstract machine does not need to know the arity of a function from its closure.)

This way, given a closure value `v`, the field number where the environment starts is the "start-of-environment" bits of `Field(v, 1)`.  

Example: native-code, two mutually recursive functions of arities 2 and 1, two free variables:
```
code1  {arity=2,start-env=6}  code1'  infix-hdr  code2  {arity=1,start_env=2}  freevar1  freevar2
```

## The pull request

It is best reviewed commit by commit:
- 9f59d7f5e changes the representation of closures in the bytecode system and 8e97e63b3 does the same for the native-code compiler.
- cd2ed2b1e shows how the GC can use the new "start of environment" information to skip over the entrypoints parts of closures.  Currently it's done only in no-naked-pointers mode, but would work in naked-pointers mode too.
- 3d1a4112c and 95bd98216 further simplify the GC in no-naked-pointers mode.
- 14c96d1ef tweaks a silly test and 7fbb5567d is the obligatory bootstrap when bytecode compilation changes.

## FAQ

To speed up reviewing, here are some questions and answers.

Q: What are the differences with #1156 ?

A: #1156 used a single bit in the arity words to mark the last entrypoint in the sequence of entrypoints.  Consequently, jumping to the environment part required scanning the entry points one after the other.  This PR turns this into a constant-time operation.  (As suggested by @chambart in https://github.com/ocaml/ocaml/pull/1156#pullrequestreview-35788965 .)

Q: What are the differences with #8984 ?

A: #8984 explores another closure representation strategy, where each function has its own heap block, containing only one entry point.  This is a much bigger change in the compilers and the run-time system, with potential for bigger simplifications (getting rid of infix pointers entirely), but the various representation choices still need to be explored and benchmarked, so it will not be ready immediately.  The present PR is the quick, minimal change that suffices for Multicore OCaml.

Q: What is the impact on execution time (GC not included)?

A: In native code, negligible: one extra shift instruction when applying an unknown function to 2 or more arguments.  In bytecode, closures become bigger and may take a little longer to create.

Q: What is the impact on GC time?

A: Currently, nothing changes in naked-pointers mode.  More generally, skipping over the entrypoints part of closures rather than scanning them might give a small speedup for the major GC.

Q: Why not steal bits in infix headers to store the "start of environment" info?

A: There is room there, but for one function we don't have an infix header.  (N-1 infix headers for a N-function closure.)

Q: Was it absolutely necessary to increase the size of bytecode closures?

A: I could not find a nice solution that has only one code pointer per entry point and no closure info field.  Plus, having similar closure representations in bytecode and in native-code might simplify other things in the future.

Q: Why only 8 bits for the arity?

A: So that it leaves enough bits for the "start of environment" field to represent any valid position in any heap block, even in 32 bit machines.  Plus, it's not a limitation in practice.

Q: But what happens if I have a curried function of 300 arguments?

A: Right now, an assertion failure at compile-time.  With #9620, your function will be compiled to a 126-argument function returning a 126-argument function returning a 48-argument function.

Q: Why the 8 most significant bits for the arity and not bits 1 to 8?

A: The arity is a signed integer (negative arity = tupled function), so a single arithmetic shift right extracts the arity with the correct sign.

Q: How far are we from a "no-naked-pointer" mode that does not need the page table at all?

A: In native code I think we're almost there.  In bytecode, some more work remains (code pointers are used as naked-pointer values in many places).

xavierleroy committed:
Give preallocated atoms the GC color Black

Atoms are zero-sized blocks allocated outside the heap.
It simplifies the GC in no-naked-pointers mode if their headers
have GC color Black, meaning "don't traverse".
In ocamlopt, Black is already used as the color for constant blocks
statically allocated outside the heap.

xavierleroy committed:
No need to special-case zero-sized blocks in no-naked-pointers mode

Now that atoms have black headers, all zero-sized blocks (atoms or
ocamlopt-generated static data) have black headers and will not
be traversed or changed by the major GC.

xavierleroy committed:
New representation of closures, bytecode part

Add a "closure information" field after each code pointer in
a closure.  This field generalizes the "arity" field used by
the native-code compiler, in that it has room both for an arity
(always 0 in bytecode) and for the distance from the closure
to the first environment variable in the closure block.

This makes closures "self-described" and easy to scan for pointers:
everything up to the first env var is out-of-heap code pointers or
integers; everything after the first env var is a well-formed value.

At this point a bootstrap is needed.

kayceesrk commented:
> Q: How far are we from a "no-naked-pointer" mode that does not need the page table at all?
> A: In native code I think we're almost there. 

Could you mention where the page table is still needed in no-naked-pointer mode?

xavierleroy commented:
> Could you mention where the page table is still needed in no-naked-pointer mode?

The things you make me do...  I checked the remaining calls to ` caml_page_table_lookup` in the native-code runtime system compiled in no-naked-pointers mode.  Here they are:
```
array.c
    function caml_make_vect: Is_in_value_area
    function caml_make_array: Is_in_value_area

compact.c
    function invert_pointer_at: Is_in_heap

extern.c
    function extern_rec: Is_in_value_area

finalise.c
    function generic_final_register: Is_in_heap_or_young
    + lots of assertions

globroots.c
    function classify_gc_root: Is_in_heap

hash.c
    function caml_hash: Is_in_value_area
    function hash_aux: Is_in_value_area

major_gc.c
   function mark_ephe_aux: Is_in_heap

minor_gc.c
    function caml_oldify_one: Is_in_value_area

obj.c
    function caml_obj_tag: Is_in_value_area
    function caml_lazy_follow_forward: Is_in_value_area
    function caml_obj_reachable_words: Is_in_heap_or_young

signals_nat.c
   function handle_signal: Is_in_code_area

caml/weak.h
   function caml_ephe_clean_partial: Is_in_heap_or_young
```

`Is_in_value_area` can be replaced with `1` in no-naked-pointers mode, if I'm not mistaken.

`Is_in_code_area` could be implemented differently, using the table of code fragments.  It is not performance-critical.

The remaining occurrences of `Is_in_heap` and `Is_in_heap_or_young` need scrutiny.

mshinwell commented on asmcomp/cmm_helpers.mli:
```diff
@@ -64,11 +64,15 @@ val boxedint32_header : nativeint
 val alloc_floatarray_header : int -> Debuginfo.t -> expression
 val alloc_closure_header : int -> Debuginfo.t -> expression
 val alloc_infix_header : int -> Debuginfo.t -> expression
+val alloc_closure_info : int -> int -> Debuginfo.t -> expression
```
Please could we label the arguments to `closure_info` and `alloc_closure_info`?

xavierleroy replied:
Okay.  See commit e6533868f .

mshinwell commented on asmcomp/cmm_helpers.ml:
```diff
@@ -72,11 +72,21 @@ let caml_nativeint_ops = "caml_nativeint_ops"
+  assert (0 <= delta && delta < 1 lsl (pos_arity_in_closinfo - 1));
+  Nativeint.(add (shift_left (of_int arity) pos_arity_in_closinfo)
+                 (add (shift_left (of_int delta) 1)
+                      1n))
```
I believe this calculation is correct even if the arity is negative (which means its top bit overlaps the sign bit of the whole closure info word).  I thought I'd point this out explicitly though.

xavierleroy replied:
Documentation is in the code.  The two `assert` above characterize exactly the allowed ranges for the arguments, including the fact that arity can be negative.

mshinwell commented on runtime/major_gc.c:
```diff
@@ -398,6 +393,13 @@ static void mark_slice (intnat work)
+      if (Tag_val(v) == Closure_tag) {
+        /* Skip the code pointers and integers at beginning of closure;
+           start scanning at the first word of the environment part. */
+        start = Start_env_closinfo(Closinfo_val(v));
```
Maybe we could have an assertion that `Start_env_closinfo` does indeed lie within the value `v`?

xavierleroy replied:
All right.  See commit c7a226436 .

mshinwell reviewed (approved):
Very good.  I checked all of the arithmetic and couldn't find any mistakes.  I left a couple of minor comments.

Regarding removal of the page table entirely, I don't think we can do this on single core OCaml, because of the presence of the compactor.

mshinwell commented:
I believe this is ready to merge once the CI passes.  Thanks.

damiendoligez reviewed (approved):
I have reviewed this PR and everything looks good. Note that there is another place where we examine pointers and could avoid looking at the code pointers of closures: the compactor. But we still need the page table there anyway.

@xavierleroy you'll have to rebase to get CI to run.

xavierleroy commented:
The rebase is a pain because of the bootstrap in the middle.  I'm running a round of CI precheck right now, while I figure out what to do.

xavierleroy committed:
Changes entry for #9619

xavierleroy committed:
New representation of closures, intermediate bootstrap

ocamlc's computations of offsets in closures changed to match changes in the
abstract machine.

xavierleroy committed:
New representation of closures, native-code compilation

In code that builds closures, instead of the old arity field,
produce a closure information field encoding arity + position of environment.

xavierleroy committed:
Restrict 'test_locations' to 64-bit archs and update expected outputs

Expected outputs contain integer values for the "closure info"
field of some closures.  These values differ in 32 and 64 bits,
since the arity is stored in top 8 bits.  This test would need
different expected outputs for 32- and 64-bit platforms.

To keep things simple, this commit restricts the test
to only run on 64-bit platforms.  Since this changes the locations

xavierleroy committed:
During major GC, scan the environment part of closures only

Here we start reaping the benefits of the new closure representation.

The fields of a closure block that contain the code pointers need not
be scanned (in general) and must not be scanned (in no-naked-pointers mode).

Here, conservatively, we skip them in no-naked-pointers mode only,
but it would be sound to skip them unconditionally.

xavierleroy commented:
CI precheck: OK.  Rebase from hell: done.  Waiting for Travis&Appveyor.

xavierleroy commented:
Kicking Travis.

xavierleroy commented:
Travis is not run, and Appveyor is pending as usual, so I'll just merge and take a well-deserved week-end.

jhjourdan commented:
I was also caught up by the bug found in #9681, which led me to do a `grep Closure_tag` in the runtime sources: apparently `caml_alloc_dummy_infix` still allocates a block with this tag, but without taking care of the closinfo field. Is this intended?

It turns out that the closinfo field is initialized by `caml_alloc` with `Val_unit`, which (by chance?) corresponds to a "somewhat valid" value of the closinfo field (if my understanding is good). If this is intended, then a comment would at the very welcome!

xavierleroy commented:
Re `caml_alloc_dummy_infix`: no, it's not intended, thanks for having spotted it.  I also need to check `caml_update_dummy` carefully.

xavierleroy commented:
> Could you mention where the page table is still needed in no-naked-pointer mode?

Once #9698 is merged, only the compactor will need the page table, more precisely this line of the compactor:
https://github.com/ocaml/ocaml/blob/5f4542862f384a63a988db1dbd49c1e6b1ebf217/runtime/compact.c#L77

Kakadu commented:
@xavierleroy Do we already have a documentation entry for new closure representation?

I'm not interested in the documentation per se, actually I want to try to implement memoization/hashconsing of partially applied high-order functions. It seems that your changes are directly related to un-(partial applying) because 

> Given a closure block at run-time, there is no way to determine which well-formed OCaml values it contains. More precisely, there is no way to know where the entrypoints end and where the environment starts, since the first value of the environment can be a tagged integer that looks like an infix header.

Last time when  I tried to dig into it, I realized that the representation changes when `-flambda` is on, and I couldn't to understand it from documentation too.

xavierleroy commented:
> Do we already have a documentation entry for new closure representation?

For the moment, there's the description at the top of this PR #9619 and you can see examples of how to traverse closures here for example: https://github.com/ocaml/ocaml/blob/302d735ce8f965f48784b90af11aaeddf22f88d1/runtime/hash.c#L264-L283

> It seems that your changes are directly related to un-(partial applying)

Yes and no.  Yes insofar as partial applications produce "interesting" closures, with non-trivial environments.  No because the change of representation applies to all closures, not just those arising out of partial applications.
